### PR TITLE
Issue/1388 slices

### PR DESCRIPTION
**Pull request by bartv**
Refactor of server in many smaller slices (called services). I also did quite some cleanup to make use of mypy to verify whether the refactor was correct. This also fixed some smaller bugs.

Up to the reviewer to decide how far to go into the rabbit hole.

- [x] unit tests
- [x] lsm tests -- 1399 is required 
- [x] performance tests -- 1399 is required
- [x] verfiy if all deps in services are still in use

### CODE REVIEW

File: src/inmanta/config.py [hunk 1]
```diff
@@ -290,7 +290,7 @@ def get_default_desc(self) -> str:
     def validate(self, value: str) -> T:
         return self.validator(value)
 
-    def get_default_value(self):
+    def get_default_value(self) -> Union[str, bool, int, None, Callable]:
```
Comment (wouterdb) [inline src/inmanta/config.py]:
```suggestion
    def get_default_value(self) -> Union[str, bool, int, None]:
```

File: src/inmanta/data/model.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ class ExtensionStatus(BaseModel):
 
 class SliceStatus(BaseModel):
     """
-        Status response for slices loaded in the the server
+        Status response for services loaded in the the server
```
Comment (wouterdb) [inline src/inmanta/data/model.py]:
```suggestion
        Status response for slices loaded in the the server
```

File: src/inmanta/server/bootloader.py [hunk 3]
```diff
@@ -60,8 +60,8 @@ class InmantaBootloader(object):
     """ The inmanta bootloader is responsible for:
         - discovering extensions
         - loading extensions
-        - loading core and extension slices
-        - starting the server and its slices in the correct order
+        - loading core and extension services
```
Comment (wouterdb) [inline src/inmanta/server/bootloader.py]:
```suggestion
        - loading core and extension slices
```

File: src/inmanta/server/bootloader.py [hunk 4]
```diff
@@ -60,8 +60,8 @@ class InmantaBootloader(object):
...
+        - starting the server and its services in the correct order
```
Comment (wouterdb) [inline src/inmanta/server/bootloader.py]:
```suggestion
        - starting the server and its slices in the correct order
```

File: src/inmanta/server/bootloader.py [hunk 5]
```diff
@@ -115,7 +115,7 @@ def _discover_plugin_packages(self) -> List[str]:
 
     def _load_extension(self, name: str) -> Callable[[ApplicationContext], None]:
         """ Import the extension defined in the package in name and return the setup function that needs to be called for the
-            extension to register its slices in the application context.
+            extension to register its services in the application context.
```
Comment (wouterdb) [inline src/inmanta/server/bootloader.py]:
```suggestion
            extension to register its slices in the application context.
```

File: src/inmanta/server/bootloader.py [hunk 6]
```diff
@@ -142,11 +142,10 @@ def _load_extensions(self) -> Dict[str, Callable[[ApplicationContext], None]]:
                 LOGGER.warning("Could not load extension %s", name, exc_info=True)
         return plugins
 
-    # Extension loading Phase II: collect slices
+    # Extension loading Phase II: collect services
```
Comment (wouterdb) [inline src/inmanta/server/bootloader.py]:
```suggestion
    # Extension loading Phase II: collect slices
```

File: src/inmanta/server/bootloader.py [hunk 7]
```diff
@@ -142,11 +142,10 @@ def _load_extensions(self) -> Dict[str, Callable[[ApplicationContext], None]]:
...
     def _collect_slices(self, extensions: Dict[str, Callable[[ApplicationContext], None]]) -> ApplicationContext:
         """
-            Call the setup function on all extensions and let them register their slices in the ApplicationContext. Core slices
-            are statically
+            Call the setup function on all extensions and let them register their services in the ApplicationContext.
```
Comment (wouterdb) [inline src/inmanta/server/bootloader.py]:
```suggestion
            Call the setup function on all extensions and let them register their slices in the ApplicationContext.
```

File: src/inmanta/server/bootloader.py [hunk 8]
```diff
@@ -156,7 +155,7 @@ def _collect_slices(self, extensions: Dict[str, Callable[[ApplicationContext], N
 
     def load_slices(self) -> List[ServerSlice]:
         """
-            Load all slices in the server
+            Load all services in the server
```
Comment (wouterdb) [inline src/inmanta/server/bootloader.py]:
```suggestion
            Load all slices in the server
```

File: src/inmanta/server/protocol.py [hunk 9]
```diff
@@ -122,7 +122,7 @@ def _order_slices(self) -> List["ServerSlice"]:
         try:
             order = stable_depth_first(names, {k: list(v) for k, v in edges.items()})
         except CycleException as e:
-            raise ServerStartFailure("Dependency cycle between server slices " + ",".join(e.nodes)) from e
+            raise ServerStartFailure("Dependency cycle between server services " + ",".join(e.nodes)) from e
```
Comment (wouterdb) [inline src/inmanta/server/protocol.py]:
```suggestion
            raise ServerStartFailure("Dependency cycle between server slices " + ",".join(e.nodes)) from e
```

File: src/inmanta/server/protocol.py [hunk 10]
```diff
@@ -219,13 +219,13 @@ def __init__(self, name: str) -> None:
...
 
-    def is_stopping(self):
+    def is_stopping(self) -> bool:
         """True when prestop has been called."""
         return self._stopping
 
     async def prestart(self, server: Server) -> None:
         """
-        Called by the RestServer host prior to start, can be used to collect references to other server slices
+        Called by the RestServer host prior to start, can be used to collect references to other server services
```
Comment (wouterdb) [inline src/inmanta/server/protocol.py]:
```suggestion
        Called by the RestServer host prior to start, can be used to collect references to other server slices
```

File: src/inmanta/server/protocol.py [hunk 11]
```diff
@@ -269,11 +269,11 @@ def is_stopping(self):
         await super(ServerSlice, self).stop()
 
     def get_dependencies(self) -> List[str]:
-        """List of names of slices that must be started before this one."""
+        """List of names of services that must be started before this one."""
```
Comment (wouterdb) [inline src/inmanta/server/protocol.py]:
```suggestion
        """List of names of slices that must be started before this one."""
```

File: src/inmanta/server/protocol.py [hunk 12]
```diff
@@ -269,11 +269,11 @@ def is_stopping(self):
...
         return []
 
     def get_depended_by(self) -> List[str]:
-        """List of names of slices that must be started after this one."""
+        """List of names of services that must be started after this one."""
```
Comment (wouterdb) [inline src/inmanta/server/protocol.py]:
```suggestion
        """List of names of slices that must be started after this one."""
```

File: src/inmanta/server/server.py [hunk 13]
```diff
@@ -188,116 +72,8 @@ def get_depended_by(self) -> List[str]:
...
-        elif len(resource_ids) > 1:
-            message = "multiple resources: " + message
-        else:
-            message = resource_ids[0] + ": " + message
-        log_record = ResourceActionLogLine(logger.name, log_level, message, ts)
-        logger.handle(log_record)
-
-    def get_agent_client(self, tid: UUID, endpoint: str) -> Optional[protocol.ReturnClient]:
-        return self.agentmanager.get_agent_client(tid, endpoint)
+        self.agentmanager: "AgentManager" = cast("AgentManager", server.get_slice(SLICE_AGENT_MANAGER))
```
Comment (arnaudsjs) [inline src/inmanta/server/server.py]: The SLICE_AGENT_MANAGER and the SLICE_COMPILER service are used here, but they are not part of the dependency list.

File: src/inmanta/server/services/environmentservice.py [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+    def __init__(self) -> None:
+        super(EnvironmentService, self).__init__(SLICE_ENVIRONMENT)
+
+    def get_dependencies(self) -> List[str]:
+        return [SLICE_SERVER, SLICE_DATABASE, SLICE_AGENT_MANAGER]
+
+    async def prestart(self, server: protocol.Server) -> None:
+        self.server_slice = cast(Server, server.get_slice(SLICE_SERVER))
+        self.agentmanager = cast(AgentManager, server.get_slice(SLICE_AGENT_MANAGER))
+        self.orchestration_service = cast(OrchestrationService, server.get_slice(SLICE_ORCHESTRATION))
```
Comment (arnaudsjs) [inline src/inmanta/server/services/environmentservice.py]: SLICE_ORCHESTRATION and SLICE_RESOURCE are used here but they are not part of the dependency list

File: src/inmanta/server/services/projectservice.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    resource_service: ResourceService
+
+    def __init__(self) -> None:
+        super(ProjectService, self).__init__(SLICE_PROJECT)
+
+    def get_dependencies(self) -> List[str]:
+        return [SLICE_DATABASE, SLICE_RESOURCE]
+
+    async def prestart(self, server: protocol.Server) -> None:
+        self.agentmanager = cast(AgentManager, server.get_slice(SLICE_AGENT_MANAGER))
```
Comment (arnaudsjs) [inline src/inmanta/server/services/projectservice.py]: SLICE_AGENT_MANAGER is not in dependency list.

File: src/inmanta/server/services/environmentservice.py [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+        self.resource_service = cast(ResourceService, server.get_slice(SLICE_RESOURCE))
+
+    # Project handlers
+    @protocol.handle(methods.create_project)
+    async def create_project(self, name: str, project_id: Optional[uuid.UUID]) -> Apireturn:
```
Comment (arnaudsjs) [inline src/inmanta/server/services/environmentservice.py]: Project handlers are already defined in the project service. It should be impossible to register two handler for the same endpoint.

**Comment by arnaudsjs** [comment]
General remarks (I didn't add a comment to every occurrence):
* The prestart(), start(), prestop() and stop() methods of a server slide should always call the same method in the superclass.
* If a slice defines an API endpoint, to should have SLICE_TRANSPORT as a depended_by

File: src/inmanta/server/services/projectservice.py (hunk 15 above)
Comment (bartv) [reply src/inmanta/server/services/projectservice.py]: done

File: src/inmanta/server/services/environmentservice.py (hunk 14 above)
Comment (bartv) [reply src/inmanta/server/services/environmentservice.py]: done

File: src/inmanta/server/server.py (hunk 13 above)
Comment (bartv) [reply src/inmanta/server/server.py]: done

File: src/inmanta/server/__init__.py [hunk 17]
```diff
@@ -24,3 +24,13 @@
 SLICE_DATABASE = "core.database"
 SLICE_TRANSPORT = "core.transport"
 SLICE_COMPILER = "core.compiler"
+SLICE_FORM = "core.forms"
+SLICE_PROJECT = "core.project"
+SLICE_ENVIRONMENT = "core.environment"
+SLICE_FILE = "core.slice"
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/server/__init__.py]: done

File: src/inmanta/server/services/codeservice.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+
+    file_slice: FileService
+
+    def __init__(self) -> None:
+        super(CodeService, self).__init__(SLICE_CODE)
+
+    def get_dependencies(self) -> List[str]:
+        return [SLICE_FILE, SLICE_DATABASE]
+
+    async def prestart(self, server: protocol.Server) -> None:
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/server/services/codeservice.py]: done

File: src/inmanta/server/services/codeservice.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+from inmanta.protocol.exceptions import BadRequest, NotFound, ServerError
+from inmanta.server import SLICE_CODE, SLICE_DATABASE, SLICE_FILE, protocol
+from inmanta.server.services.fileservice import FileService
+from inmanta.types import Apireturn, JsonType
+from inmanta.util import hash_file
+
+LOGGER = logging.getLogger(__name__)
+
+
+class CodeService(protocol.ServerSlice):
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/server/services/codeservice.py]: done

File: src/inmanta/server/services/environmentservice.py (hunk 16 above)
Comment (bartv) [reply src/inmanta/server/services/environmentservice.py]: done

**Pull request merged**